#  Importing necessary libraries and packages

In [1]:
import pandas as pd # library for data analsysis
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
from bs4 import BeautifulSoup
 

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab3
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [2]:
# @hidden_cell
CLIENT_ID = '1PWQHF2E3QY5EX4JN2WVVC1D3URZL4KS1W3K1RUQZAOPUXJJ' # your Foursquare ID
CLIENT_SECRET = 'UP352PBZMNM0VP23OTEM0TD1LLL53UCH04FPAHQYVPMIPVBZ' # your Foursquare Secret


In [3]:
VERSION = '20180604'
LIMIT = 30

### Webscraping from WIKIpeadia

### URL

In [4]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
url

'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

### HTML stored in resuls

In [5]:
#  webscrape data from wikipedia and get the html
results = requests.get(url).text
results

'\n<!DOCTYPE html>\n<html class="client-nojs" lang="en" dir="ltr">\n<head>\n<meta charset="UTF-8"/>\n<title>List of postal codes of Canada: M - Wikipedia</title>\n<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"6e2ff215-e017-4593-98e0-790d52e9651e","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969510799,"wgRevisionId":969510799,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toronto","Ontar

### Results conversion into  XML and creation of table 

In [6]:

Torontosoup = BeautifulSoup(results, 'xml')
Torontotable=Torontosoup.find('table')
Torontosoup

<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="UTF-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"6e2ff215-e017-4593-98e0-790d52e9651e","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969510799,"wgRevisionId":969510799,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal co

### Creating the empty dataframe for load of table

In [7]:
# empty dataframe
column_names=['Postalcode','Borough','Neighborhood']
dfToronto = pd.DataFrame(columns=column_names)
dfToronto

Empty DataFrame
Columns: [Postalcode, Borough, Neighborhood]
Index: []

### Read table and append in dataframe

In [8]:
# load the results of the Torontotable in dataframe
 

for row in Torontotable.find_all('tr'):
    row_data=[]
    for cell in row.find_all('td'):
        row_data.append(cell.text.strip())
    if len(row_data)==3:
        dfToronto.loc[len(dfToronto)] = row_data
dfToronto.shape


(180, 3)

# 180 Records in dataframe

In [9]:
dfToronto['Borough'].value_counts()
 

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East Toronto         5
York                 5
East York            5
Mississauga          1
Name: Borough, dtype: int64

### Total 180 records and 77 records with Borough = Not assigned

In [10]:
 
dfToronto.loc[dfToronto['Borough'] == 'Not assigned']
 

Postalcode       Borough  Neighborhood
0          M1A  Not assigned  Not assigned
1          M2A  Not assigned  Not assigned
7          M8A  Not assigned  Not assigned
10         M2B  Not assigned  Not assigned
15         M7B  Not assigned  Not assigned
16         M8B  Not assigned  Not assigned
19         M2C  Not assigned  Not assigned
24         M7C  Not assigned  Not assigned
25         M8C  Not assigned  Not assigned
28         M2E  Not assigned  Not assigned
29         M3E  Not assigned  Not assigned
33         M7E  Not assigned  Not assigned
34         M8E  Not assigned  Not assigned
35         M9E  Not assigned  Not assigned
37         M2G  Not assigned  Not assigned
38         M3G  Not assigned  Not assigned
42         M7G  Not assigned  Not assigned
43         M8G  Not assigned  Not assigned
44         M9G  Not assigned  Not assigned
51         M7H  Not assigned  Not assigned
52         M8H  Not assigned  Not assigned
53         M9H  Not assigned  Not assigned
60         M7J  Not assigned  Not assigned
61         M8J  Not assigned  Not assigned
62         M9J  Not assigned  Not assigned
69         M7K  Not assigned  Not assigned
70         M8K  Not assigned  Not assigned
71         M9K  Not assigned  Not assigned
78         M7L  Not assigned  Not assigned
79         M8L  Not assigned  Not assigned
87         M7M  Not assigned  Not assigned
88         M8M  Not assigned  Not assigned
96         M7N  Not assigned  Not assigned
97         M8N  Not assigned  Not assigned
101        M3P  Not assigned  Not assigned
105        M7P  Not assigned  Not assigned
106        M8P  Not assigned  Not assigned
110        M3R  Not assigned  Not assigned
115        M8R  Not assigned  Not assigned
118        M2S  Not assigned  Not assigned
119        M3S  Not assigned  Not assigned
123        M7S  Not assigned  Not assigned
124        M8S  Not assigned  Not assigned
125        M9S  Not assigned  Not assigned
127        M2T  Not assigned  Not assigned
128        M3T  Not assigned  Not assigned
131        M6T  Not assigned  Not assigned
132        M7T  Not assigned  Not assigned
133        M8T  Not assigned  Not assigned
134        M9T  Not assigned  Not assigned
136        M2V  Not assigned  Not assigned
137        M3V  Not assigned  Not assigned
140        M6V  Not assigned  Not assigned
141        M7V  Not assigned  Not assigned
145        M2W  Not assigned  Not assigned
146        M3W  Not assigned  Not assigned
149        M6W  Not assigned  Not assigned
150        M7W  Not assigned  Not assigned
154        M2X  Not assigned  Not assigned
155        M3X  Not assigned  Not assigned
158        M6X  Not assigned  Not assigned
159        M7X  Not assigned  Not assigned
161        M9X  Not assigned  Not assigned
162        M1Y  Not assigned  Not assigned
163        M2Y  Not assigned  Not assigned
164        M3Y  Not assigned  Not assigned
166        M5Y  Not assigned  Not assigned
167        M6Y  Not assigned  Not assigned
170        M9Y  Not assigned  Not assigned
171        M1Z  Not assigned  Not assigned
172        M2Z  Not assigned  Not assigned
173        M3Z  Not assigned  Not assigned
174        M4Z  Not assigned  Not assigned
175        M5Z  Not assigned  Not assigned
176        M6Z  Not assigned  Not assigned
177        M7Z  Not assigned  Not assigned
179        M9Z  Not assigned  Not assigned

In [11]:
#Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
dfToronto=dfToronto[dfToronto['Borough']!='Not assigned']
dfToronto.loc[dfToronto['Borough'] == 'Not assigned']


Empty DataFrame
Columns: [Postalcode, Borough, Neighborhood]
Index: []

In [12]:
dfToronto.shape

(103, 3)

### Hence out of 180 records 77 records deleted, left with 103 records

### If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
 
 


In [13]:
dfToronto.loc[dfToronto['Neighborhood'] == 'Not assigned', 'Neighborhood'] = dfToronto['Borough']
dfToronto.shape

D:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)
D:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


(103, 3)

### Combining rows for a single postcal code

### Let us see the unique count 

In [14]:

postalcodes = dfToronto['Postalcode'].nunique()
boroughs = dfToronto['Borough'].nunique()
neighbourhoods= dfToronto['Neighborhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

Unique Postalcodes : 103
Unique Boroughs  : 10
Unique Neighbourhoods  :99


### Note : The wikipedia site has changed and now contains combined nieghbourhoods , however code is provided

In [15]:
# group multiple Neighbourhood under one Postcode
temp=dfToronto.groupby('Postalcode')['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
temp=temp.reset_index(drop=False)
temp.rename(columns={'Neighborhood':'Neighborhood_joined'},inplace=True)
temp

Postalcode                                Neighborhood_joined
0          M1B                                     Malvern, Rouge
1          M1C             Rouge Hill, Port Union, Highland Creek
2          M1E                  Guildwood, Morningside, West Hill
3          M1G                                             Woburn
4          M1H                                          Cedarbrae
5          M1J                                Scarborough Village
6          M1K        Kennedy Park, Ionview, East Birchmount Park
7          M1L                    Golden Mile, Clairlea, Oakridge
8          M1M    Cliffside, Cliffcrest, Scarborough Village West
9          M1N                        Birch Cliff, Cliffside West
10         M1P  Dorset Park, Wexford Heights, Scarborough Town...
11         M1R                                  Wexford, Maryvale
12         M1S                                          Agincourt
13         M1T            Clarks Corners, Tam O'Shanter, Sullivan
14         M1V  Milliken, Agincourt North, Steeles East, L'Amo...
15         M1W                      Steeles West, L'Amoreaux West
16         M1X                                        Upper Rouge
17         M2H                                  Hillcrest Village
18         M2J                       Fairview, Henry Farm, Oriole
19         M2K                                    Bayview Village
20         M2L                           York Mills, Silver Hills
21         M2M                            Willowdale, Newtonbrook
22         M2N                        Willowdale, Willowdale East
23         M2P                                    York Mills West
24         M2R                        Willowdale, Willowdale West
25         M3A                                          Parkwoods
26         M3B                                          Don Mills
27         M3C                                          Don Mills
28         M3H    Bathurst Manor, Wilson Heights, Downsview North
29         M3J                    Northwood Park, York University
30         M3K                                          Downsview
31         M3L                                          Downsview
32         M3M                                          Downsview
33         M3N                                          Downsview
34         M4A                                   Victoria Village
35         M4B                    Parkview Hill, Woodbine Gardens
36         M4C                                   Woodbine Heights
37         M4E                                        The Beaches
38         M4G                                            Leaside
39         M4H                                   Thorncliffe Park
40         M4J      East Toronto, Broadview North (Old East York)
41         M4K                       The Danforth West, Riverdale
42         M4L                     India Bazaar, The Beaches West
43         M4M                                    Studio District
44         M4N                                      Lawrence Park
45         M4P                                   Davisville North
46         M4R                 North Toronto West,  Lawrence Park
47         M4S                                         Davisville
48         M4T                        Moore Park, Summerhill East
49         M4V  Summerhill West, Rathnelly, South Hill, Forest...
50         M4W                                           Rosedale
51         M4X                        St. James Town, Cabbagetown
52         M4Y                               Church and Wellesley
53         M5A                          Regent Park, Harbourfront
54         M5B                           Garden District, Ryerson
55         M5C                                     St. James Town
56         M5E                                        Berczy Park
57         M5G                                 Central Bay Street
58         M5H                           Richmond, Adelaide, King
59         M5J  Harbourfront East, Union Sta

In [16]:
df_mergeToronto = pd.merge(dfToronto, temp, on='Postalcode')
df_mergeToronto

Postalcode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  \
0                                            Parkwoods   
1                                     Victoria Village   
2                            Regent Park, Harbourfront   
3                     Lawre

In [17]:
# drop the Neighbourhood column
df_mergeToronto.drop(['Neighborhood'],axis=1,inplace=True)

# drop duplicates from the data frame
df_mergeToronto.drop_duplicates(inplace=True)

# rename Neighbourhood_joined back to Neighbourhood
df_mergeToronto.rename(columns={'Neighborhood_joined':'Neighborhood'},inplace=True)
df_mergeToronto.shape

(103, 3)

### Creation of data frame is completed

### Loading Geographical co-ordinates

In [18]:
location=pd.read_csv('http://cocl.us/Geospatial_data')
location

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

### Merging the toronto dataframe and location : cordinates dataframe and correction of order

In [19]:
location.rename(columns={'Postal Code':'Postalcode'},inplace=True)
geoToronto = pd.merge(location, df_mergeToronto, on='Postalcode')
# correcting the sequence of data
geoToronto=geoToronto[['Postalcode','Borough','Neighborhood','Latitude','Longitude']]
geoToronto

Postalcode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    G

In [20]:
postalcodes = geoToronto['Postalcode'].nunique()
boroughs = geoToronto['Borough'].nunique()
neighbourhoods= geoToronto['Neighborhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

Unique Postalcodes : 103
Unique Boroughs  : 10
Unique Neighbourhoods  :99


In [21]:
# only boroughs that contain the word Toronto 
dfToronto=geoToronto[geoToronto['Borough'].str.contains("Toronto")]
dfToronto.reset_index(drop=False, inplace=True)
dfToronto

index Postalcode           Borough  \
0      37        M4E      East Toronto   
1      41        M4K      East Toronto   
2      42        M4L      East Toronto   
3      43        M4M      East Toronto   
4      44        M4N   Central Toronto   
5      45        M4P   Central Toronto   
6      46        M4R   Central Toronto   
7      47        M4S   Central Toronto   
8      48        M4T   Central Toronto   
9      49        M4V   Central Toronto   
10     50        M4W  Downtown Toronto   
11     51        M4X  Downtown Toronto   
12     52        M4Y  Downtown Toronto   
13     53        M5A  Downtown Toronto   
14     54        M5B  Downtown Toronto   
15     55        M5C  Downtown Toronto   
16     56        M5E  Downtown Toronto   
17     57        M5G  Downtown Toronto   
18     58        M5H  Downtown Toronto   
19     59        M5J  Downtown Toronto   
20     60        M5K  Downtown Toronto   
21     61        M5L  Downtown Toronto   
22     63        M5N   Central Toronto   
23     64        M5P   Central Toronto   
24     65        M5R   Central Toronto   
25     66        M5S  Downtown Toronto   
26     67        M5T  Downtown Toronto   
27     68        M5V  Downtown Toronto   
28     69        M5W  Downtown Toronto   
29     70        M5X  Downtown Toronto   
30     75        M6G  Downtown Toronto   
31     76        M6H      West Toronto   
32     77        M6J      West Toronto   
33     78        M6K      West Toronto   
34     82        M6P      West Toronto   
35     83        M6R      West Toronto   
36     84        M6S      West Toronto   
37     85        M7A  Downtown Toronto   
38     87        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                  North Toronto West,  Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                                     Stn A PO Boxes  43.646

In [22]:
dfToronto.shape

(39, 6)

### This completes part 1 of the project

# Start of Part 2

In [23]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [24]:
# create map of New York using latitude and longitude values
map = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(dfToronto['Latitude'], dfToronto['Longitude'], dfToronto['Borough'], dfToronto['Neighborhood']):
    label = '{}, {}'.format(dfToronto, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map)  
    
map

### Getting venues near toronto
### Function for venues

In [25]:

def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Collecting the venues

In [26]:

toronto_venues = getNearbyVenues(names=dfToronto['Neighborhood'],
                                   latitudes=dfToronto['Latitude'],
                                   longitudes=dfToronto['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High

In [27]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [28]:
# Let's check how many venues were returned for each neighborhood

In [29]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             32   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          26   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              65   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               48   
Moore Park, Summerhill East                                             4   
North Toronto West,  Lawrence Park                                     20   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     35   
St. James Town                                                         87   
St. James Town, Cabbagetown                                            47   
Stn A PO Boxes                                                         97   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              32   
Davisville North                    

#### Let's find out how many unique categories can be curated from all the returned venues

In [30]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


### Analyze Each Neighborhood

In [31]:
# one hot encoding


toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'],axis=1,inplace=True) # seems like it already contains Neighborhood column as number
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'].to_list() )
toronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  \
0                   The Beaches                  0        0   
1                   The Beaches                  0        0   
2                   The Beaches                  0        0   
3                   The Beaches                  0        0   
4  The Danforth West, Riverdale                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Business Service  Butcher  \
0             0              0         0                 0        0   
1             0              0         0                 0        0   
2             0              0         0                 0        0   
3             0              0         0                 0        0   
4             0              0         0                 0        0   

   Cafeteria  Café  Cajun / Creole Restaurant  Cam

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [32]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013158   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                 North Toronto West,  Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Lounge  Airport Service  \
0   0.000000            0.000000        0.000000         0.000000   
1   0.000000            0.000000        0.000000         0.000000   
2   0.000000            0.000000        0.000000         0.000000   
3   0.058824            0.058824        0.117647         0.176471   
4   0.000000            0.000000        0.000000         0.000000   
5   0.000000            0.000000        0.000000         0.000000   
6   0.000000            0.000000        0.000000         0.000000   
7   0.000000            0.000000        0.000000         0.000000   
8   0.000000            0.000000        0.000000         0.000000   
9   0.000000            0.000000        0.000000         0.000000   
10  0.000000            0.000000        0.000000         0.000000   
11  0.000000            0.000000        0.000000         0.000000   
12  0.000000            0.000000        0.000000         0.000000   
13  0.000000            0.000000  

#### Let's confirm the new size

In [33]:
toronto_grouped.shape

(39, 235)

#### Let's print each neighborhood along with the top 5 most common venues

In [34]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.10
1            Café  0.03
2    Cocktail Bar  0.03
3  Farmers Market  0.03
4      Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
                venue  freq
0                Café  0.14
1      Breakfast Spot  0.09
2         Coffee Shop  0.09
3  Italian Restaurant  0.05
4           Pet Store  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2        Comic Shop  0.06
3              Park  0.06
4  Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3             Plane  0.06
4     Boat or Ferry  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1 

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Seafood Restaurant           Cheese Shop              Beer Bar   
1        Breakfast Spot           Coffee Shop             Pet Store   
2         Auto Workshop    Light Rail Station         Burrito Place   
3        Airport Lounge      Airport Terminal           Coffee Shop   
4        Sandwich Place    Italian Restaurant   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery            Restaurant                  Café   
1          Intersection                   Bar            Restaurant   
2               Butcher        Farmers Market  Fast Food Restaurant   
3       Harbor / Marina                 Plane   Rental Car Location   
4                  Café           Salad Place       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market          Cocktail Bar      Indian Restaurant  
1                Bakery    Italian Restaurant           Climbing Gym  
2            Restaurant      Recording Studio             Skate Park  
3      Sculpture Garden         Boat or Ferry                    Bar  
4          Burger Joint      Department Store            Yoga Studio

### Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [37]:
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = dfToronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged =Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() # check the last columns!

index Postalcode          Borough                    Neighborhood  \
0     37        M4E     East Toronto                     The Beaches   
1     41        M4K     East Toronto    The Danforth West, Riverdale   
2     42        M4L     East Toronto  India Bazaar, The Beaches West   
3     43        M4M     East Toronto                 Studio District   
4     44        M4N  Central Toronto                   Lawrence Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.676357 -79.293031               2                 Trail   
1  43.679557 -79.352188               2      Greek Restaurant   
2  43.668999 -79.315572               2                  Park   
3  43.659526 -79.340923               2                  Café   
4  43.728020 -79.388790               0                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub     Health Food Store        Discount Store   
1           Coffee Shop    Italian Restaurant        Ice Cream Shop   
2     Fish & Chips Shop         Movie Theater        Ice Cream Shop   
3           Coffee Shop             Gastropub               Brewery   
4    Dim Sum Restaurant              Bus Line           Swim School   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Deli / Bodega      Department Store          Dessert Shop   
1  Furniture / Home Store            Restaurant       Bubble Tea Shop   
2    Fast Food Restaurant      Sushi Restaurant               Brewery   
3                  Bakery   American Restaurant           Yoga Studio   
4             Yoga Studio         Deli / Bodega   Dumpling Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0       Dim Sum Restaurant                 Diner            Yoga Studio  
1                   Bakery                   Pub            Pizza Place  
2                      Pub            Restaurant     Italian Restaurant  
3  Comfort Food Restaurant        Sandwich Place            Cheese Shop  
4               Donut Shop      Doner Restaurant                Dog Run

Finally, let's visualize the resulting clusters

In [39]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### This completes part c of the project